src/librustc_mir/hair/pattern/_match.rs

```diff
@@ -662,11 +815,49 @@ pub fn is_useful<'p, 'a: 'p, 'tcx: 'a>(cx: &mut MatchCheckCtxt<'a, 'tcx>,
             pat_constructors(cx, row[0], pcx).unwrap_or(vec![])
         }).collect();
         debug!("used_ctors = {:#?}", used_ctors);
-        let all_ctors = all_constructors(cx, pcx);
+        // `all_ctors` are all the constructors for the given type, which
+        // should all be represented (or caught with the wild pattern `_`).
+        // `value_constructors` is true if we may exhaustively consider all
+        // the possible values (e.g. integers) of a type as its constructors.
+        let (all_ctors, value_constructors) = all_constructors(cx, pcx);
         debug!("all_ctors = {:#?}", all_ctors);
-        let missing_ctors: Vec<Constructor> = all_ctors.iter().filter(|c| {
-            !used_ctors.contains(*c)
-        }).cloned().collect();
+
+        // `missing_ctors` are those that should have appeared
+        // as patterns in the `match` expression, but did not.
+        let mut missing_ctors = vec![];
+        'req: for req_ctor in &all_ctors {
+            let mut sub_ctors = vec![req_ctor.clone()];
+            // The only constructor patterns for which it is valid to
+            // treat the values as constructors are ranges (see
+            // `all_constructors` for details).
+            let consider_value_constructors = value_constructors && match req_ctor {
+                ConstantRange(..) => true,
+                _ => false,
+            };
+            for used_ctor in &used_ctors {
+                if consider_value_constructors {
```

Comment:
If you move this `if` upwards, you don't need to introduce `sub_ctors` in the `false` case.